ocaml-lsp-server/src/hover_req.ml

```diff
@@ -446,15 +397,20 @@ let handle server { HoverParams.textDocument = { uri }; position; _ } mode =
             ~uri
             ~position
             ~with_syntax_doc
-        | Some ((`Ppx_expr _ | `Ppx_typedef_attr _) as ppx_kind) -> (
-          let+ ppx_parsetree =
-            Document.Merlin.with_pipeline_exn
-              ~name:"expand-ppx"
-              (Document.merlin_exn doc)
-              (fun pipeline -> Mpipeline.ppx_parsetree pipeline)
-          in
-          match ppx_kind with
-          | `Ppx_expr (expr, ppx) ->
-            ppx_expression_hover ~ppx_parsetree ~expr ~ppx
-          | `Ppx_typedef_attr (decl, attr) ->
-            typedef_attribute_hover ~ppx_parsetree ~decl ~attr)))
+        | Some (Ppx_expr _ | Ppx_sg_attr _ | Ppx_str_attr _) -> (
+          match
+            Document.Merlin.ppx_expand pipeline (Position.logical position)
+          with
+          | Some ppx ->
+            let contents = format_ppx_expansion ~expansion:ppx.code in
+            let range =
+              Range.create
+                ~end_:
+                  (Option.value_exn (Position.of_lexical_position ppx.attr_end))
+                ~start:
+                  (Option.value_exn
+                     (Position.of_lexical_position ppx.attr_start))
+            in
+            let hover = Hover.create ~contents ~range () in
+            Fiber.return (Some hover)
+          | None -> Fiber.return None)))
```

Comment:
Since:

1. The implementation in Merlin is already two steps:
- Determine if the node is a ppx and which kind of ppx it is
- If it is a ppx gather the parsetree items and return them

2. The current implementation is LSP uses a similar two steps process.

It feels redundant to use query dispatch and have merlin re-check the ppx node since lsp already did it.
Would it be possible to reuse the information obtained by LSP: `Some (Ppx_expr _ | Ppx_sg_attr _ | Ppx_str_attr _)` directly by calling the appropriate function in `Merlin_analysis.PPx_expand` ?

Reply:
Yes this is very possible and is indeed a better implementation.